pkg/clusterclientstore/store.go

```diff
@@ -286,14 +283,22 @@ func (s *Store) create(cluster *shipper.Cluster, secret *corev1.Secret) error {
 	}
 
 	clusterName := cluster.Name
-	newCachedCluster := cache.NewCluster(clusterName, checksum, client, config, informerFactory, func() {
-		// If/when the informer cache finishes syncing, bind all of the event handler
-		// callbacks from the controllers if it does not finish (because the cluster
-		// was Shutdown) this will not be called.
-		for _, cb := range s.eventHandlerRegisterFuncs {
-			cb(informerFactory, clusterName)
-		}
-	})
+	newCachedCluster := cache.NewCluster(
+		clusterName,
+		checksum,
+		config,
+		informerFactory,
+		func(cluster, ua string, config *rest.Config) (kubernetes.Interface, error) {
+			return s.buildClient(cluster, ua, config)
+		},
```

Comment:
I believe you can avoid this anonymous function definition and keep s.buildClient (without parentheses)(please let me know if I'm wrong).

Reply:
I did try :) It complains with the following:

```
pkg/clusterclientstore/store.go:291:4: cannot use s.buildClient (type ClientBuilderFunc) as type "github.com/bookingcom/shipper/pkg/clusterclientstore/cache".clientBuilderFunc in argument to "github.com/bookingcom/shipper/pkg/clusterclientstore/cache".NewCluster
```

I could use the `store.ClientBuilderFunc` type in the `cache` package, but the cyclical dependency doesn't sound right. Open to suggestions here.

Reply:
Oh, shame on me then. Let's keep it this way, not a big deal.

Reply:
You have two distinct named types: `cache.clientBuilderFunc` and `clusterclientstore.ClientBuilderFunc`. If you really want to keep it a part of the public API, you can put it in a separate package. If the type is deemed more internal, you can just make it not named (just put the entire signature inside `Store`).